vsphere/resource_vsphere_host.go

```diff
@@ -0,0 +1,683 @@
+	// knowing that the host does not exist any more.
+
+	// Look for host
+	client := meta.(*VSphereClient).vimClient
+	hostID := d.Id()
+	if hostID == "" {
+		return nil
+	}
+
+	// Find host and get reference to it.
+	hs, err := hostsystem.FromID(client, hostID)
+	if err != nil {
+		if soap.IsSoapFault(err) {
+			sf := soap.ToSoapFault(err)
+			_, ok := sf.Detail.Fault.(types.ManagedObjectNotFound)
+			if !ok {
+				log.Printf("[DEBUG] Error while searching host %s. Error: %s", hostID, err)
+				return err
+			}
+			d.SetId("")
+			return nil
+		}
+		log.Printf("[DEBUG] non SOAP error while searching host %s. Error %#v", hostID, err)
+		return err
+	}
+
+	maintenanceState, err := hostsystem.HostInMaintenance(hs)
+	if err != nil {
+		return err
+	}
+	d.Set("maintenance", maintenanceState)
+
+	// Retrieve host's properties.
+	log.Printf("[DEBUG] Got host %#v", hs)
+	host, err := hostsystem.Properties(hs)
+	if err != nil {
+		log.Printf("[DEBUG] Error while retrieving properties for host %s. Error: %s", hostID, err)
+		return err
+	}
+
+	if host.Parent != nil {
+		d.Set("cluster", host.Parent.Value)
+	} else {
+		d.Set("cluster", "")
+	}
+
+	connectionState, err := hostsystem.GetConnectionState(hs)
+	if err != nil {
+		return err
+	}
+
+	if connectionState == types.HostSystemConnectionStateDisconnected {
+		// Config and LicenseManager cannot be used while the host is
+		// disconnected.
+		d.Set("connected", false)
+		return nil
+	}
+	d.Set("connected", true)
+
+	lockdownMode, err := hostLockdownString(host.Config.LockdownMode)
+	if err != nil {
+		return nil
+	}
+
+	log.Printf("Setting lockdown to %s", lockdownMode)
+	d.Set("lockdown", lockdownMode)
+
+	lm := license.NewManager(client.Client)
+	am, err := lm.AssignmentManager(context.TODO())
+	if err != nil {
+		return err
+	}
+	licenses, err := am.QueryAssigned(context.TODO(), hostID)
+	if err != nil {
+		return err
+	}
+
+	licenseKey := d.Get("license").(string)
+	licFound := false
+	for _, lic := range licenses {
+		if licenseKey == lic.AssignedLicense.LicenseKey {
+			licFound = true
+			break
+		}
+	}
+
+	if !licFound {
+		d.Set("license", "")
+	}
+
+	return nil
+}
+
+func resourceVsphereHostCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*VSphereClient).vimClient
+
+	clusterID := d.Get("cluster").(string)
+	ccr, err := clustercomputeresource.FromID(client, clusterID)
+	if err != nil {
+		log.Printf("[DEBUG] Error while searching cluster %s. Error: %s", clusterID, err)
+		return err
+	}
+
+	hcs := types.HostConnectSpec{
+		HostName:      d.Get("hostname").(string),
+		UserName:      d.Get("username").(string),
+		Password:      d.Get("password").(string),
+		SslThumbprint: d.Get("thumbprint").(string),
+		Force:         d.Get("force").(bool),
+	}
+
+	licenseKey := d.Get("license").(string)
+
+	lm := license.NewManager(client.Client)
+	ll, err := lm.List(context.TODO())
+	if err != nil {
+		return err
+	}
+
+	licFound := false
```

Comment:
This could be broken out into a helper function since it is used several times.

Reply:
Unfortunately it's not quite the same. Each time I'm iterating different structures, first one is list of licences assigned to a host and the second time is a list of licenses installed on the system.

I'm concerned that getting that into a function that would support both would make the code more complicated with little benefit, but I'm open to suggestions :).

Reply:
That makes sense! I just overlooked the difference. Makes sense to keep them separate.

Reply:
For what it's worth I've ended up splitting both blocks on functions of their own, to keep the main CRUD functions small and easier (I hope) to read. 

I'll follow up with a second PR for that though :)